src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/ConsiderPassingBaseTypesAsParameters.cs

```diff
@@ -0,0 +1,227 @@
+    public sealed class ConsiderPassingBaseTypesAsParameters : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA1011";
+
+        private static readonly LocalizableString s_localizableTitle =
+            new LocalizableResourceString(
+                nameof(MicrosoftCodeQualityAnalyzersResources.ConsiderPassingBaseTypesAsParametersTitle),
+                MicrosoftCodeQualityAnalyzersResources.ResourceManager,
+                typeof(MicrosoftCodeQualityAnalyzersResources));
+
+        private static readonly LocalizableString s_localizableStandardMessage =
+            new LocalizableResourceString(
+                nameof(MicrosoftCodeQualityAnalyzersResources.ConsiderPassingBaseTypesAsParametersMessage),
+                MicrosoftCodeQualityAnalyzersResources.ResourceManager,
+                typeof(MicrosoftCodeQualityAnalyzersResources));
+
+        private static readonly LocalizableString s_localizableDescription =
+            new LocalizableResourceString(
+                nameof(MicrosoftCodeQualityAnalyzersResources.ConsiderPassingBaseTypesAsParametersDescription),
+                MicrosoftCodeQualityAnalyzersResources.ResourceManager,
+                typeof(MicrosoftCodeQualityAnalyzersResources));
+
+        internal static readonly DiagnosticDescriptor Rule =
+            DiagnosticDescriptorHelper.Create(
+                RuleId,
+                s_localizableTitle,
+                s_localizableStandardMessage,
+                DiagnosticCategory.Design,
+                RuleLevel.IdeHidden_BulkConfigurable,
+                description: s_localizableDescription,
+                isPortedFxCopRule: true,
+                isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                var eventArgsTypeSymbol = context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemEventArgs);
+
+                context.RegisterOperationBlockStartAction(context =>
+                {
+                    if (context.OwningSymbol is not IMethodSymbol methodSymbol ||
+                        methodSymbol.IsOverride ||
+                        methodSymbol.Parameters.Length == 0 ||
+                        methodSymbol.IsImplementationOfAnyInterfaceMember() ||
+                        methodSymbol.HasEventHandlerSignature(eventArgsTypeSymbol) ||
+                        !context.Options.MatchesConfiguredVisibility(Rule, methodSymbol, context.Compilation,
+                            defaultRequiredVisibility: SymbolVisibilityGroup.All) ||
+                        context.Options.IsConfiguredToSkipAnalysis(Rule, methodSymbol,
```

Comment:
Should we instead always default to reporting public APIs? The reasoning behind the rule is "allowing broader API usage", which doesn't make much sense for internal/private APIs.

Reply:
I don't know, for me it makes sense to still report on private functions if I want to because it allows for a better chance of reusability.

Reply:
@mavasani What's your opinion?

Reply:
Note that all the FxCop Design rules always operated on just the public API surface by default. So, it makes sense to default to public, but support the `api_surface` option to enable other visibilities.